runbooks/kafka-rebalance.md

```diff
@@ -0,0 +1,168 @@
+# How to Rebalance Kafka Partitions After Adding Brokers
+
+When you scale your Apache Kafka cluster by adding more broker instances, you also need to **rebalance** the partition distribution to make use of the new capacity. Kafka does **not** automatically redistribute partitions, so if you skip rebalancing, your new brokers sit idle while old ones remain overloaded.
+
+This guide walks through how to rebalance Kafka partitions using the CLI tools, including how to get the broker list from ZooKeeper.
+
+---
+
+## 📌 Why Rebalancing Is Needed
+
+- New Kafka brokers do not automatically take on any partition leadership or replicas.
+- Existing brokers continue to handle the full load.
+- Rebalancing spreads the load and improves performance and fault tolerance.
+
+---
+
+## 🔧 Step-by-Step: Kafka Rebalancing Guide
+
+### ✅ Step 0: Get the Broker List from ZooKeeper
+
+Use it if partitions are imbalanced. You can check the data directory to see if there is imbalance.
+
+Find the # of replicas and their usage
+
+```bash
+$ num_replicas=$(kubectl get statefulsets.apps kafka | grep -v "NAME" | awk '{print $2}' | awk -F'/' '{print $2}')
+
+$ for i in 0 1 2 3 4 5 6 7 8;do echo "usage for kafka-broker-${i}"; kubectl exec -it kafka-broker-${i} -- du -hd1 /bitnami/kafka/data | grep -v "__consumer" | grep -v "__transaction" ; done
+```
+
+
+### ✅ Step 1: Get the Broker List from ZooKeeper
+
+If you're using ZooKeeper (not KRaft mode), get the list of broker IDs like this:
+
+```bash
+zkCli.sh -server localhost:2181
+```
+
+Then run inside the shell:
+
+```bash
+ls /brokers/ids
+```
+
+You'll get output like:
+
+```
+[100, 101, 102, 103]
+```
+
+Use these broker IDs in your reassignment plan.
+
+Exit the shell with:
+
+```bash
+quit
+```
+
+---
+
+### ✅ Step 2: View Current Partition Distribution
+
+You can inspect how partitions are currently distributed across brokers:
+
+```bash
+kubectl exec -ti -n kfuse kafka-broker-0 -- bash
+unset JMX_PORT
+/opt/bitnami/kafka/bin/kafka-topics.sh --bootstrap-server :9092 --list
+```
+
+---
+
+### ✅ Step 3: Create a Topics JSON File
+
+You’ll need a JSON file listing the topics to rebalance. Create a file called `topics.json`:
```

Comment:
we should list out the exact steps on how to create this file in the pod.

We can either create locally and then kubectl cp

or inside the pod
run

```
cat > /bitnami/kafka/topics.json
```